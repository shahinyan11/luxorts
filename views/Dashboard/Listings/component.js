import { FormattedMessage } from 'react-intl';
import { Checkbox, Dropdown, Skeleton, Table } from 'antd';
import Link from 'next/link';
import classNames from 'classnames';

import ROUTES from 'constants/routes';
import {
  LISTING_TABLE_COLUMNS_KEY,
  LISTING_TABLE_SETTINGS,
} from 'constants/listing/manageListings';

import isPresent from 'utils/isPresent';

import DashboardLayout from 'views/layouts/Dashboard';
import GradientButton from 'views/shared/GradientButton';
import SvgIcon from 'views/shared/SvgIcon';

import Filters from './Filters';
import SortButton from './SortButton';
import EmptyState from './EmptyState';

import useContainer, { getInitialProps } from './hook';

const Listings = () => {
  const {
    total,
    isFetching,
    tableColumns,
    dataSource,
    sort,
    pagination,
    settingsVisible,
    settings,
    filters,
    searchQuery,
    amenities,
    hasAppliedFilters,
    onSortClickHandler,
    onPaginationChangeHandler,
    onSettingsChangeHandler,
    onSettingsVisibleChangeHandler,
    onFilter,
    clearAllFilters,
    onSearch,
  } = useContainer();
  const noListings = total === 0;
  const showNoResults = hasAppliedFilters || isPresent(searchQuery);

  if (isFetching) {
    return (
      <section className="listing">
        <div className="listing__header">
          <Skeleton active />
        </div>
      </section>
    );
  }

  return (
    <section className="listing">
      <div className="listing__header mb-24">
        <h1 className="listing__title">
          <FormattedMessage
            id="shared.yourListingsCounter"
            values={{ count: <span className="listing__counter">{total}</span> }}
          />
        </h1>
        {(!noListings || showNoResults) && (
          <Link href={ROUTES.ADD_NEW_LISTING.INDEX.PATH}>
            <a>
              <GradientButton text={{ id: 'shared.addNewListing' }} className="min-width-140" />
            </a>
          </Link>
        )}
      </div>
      <Filters
        amenities={amenities}
        filters={filters}
        onFilter={onFilter}
        onSearch={onSearch}
        searchQuery={searchQuery}
      />
      {noListings && (
        <EmptyState
          clearAllFilters={clearAllFilters}
          hasAppliedFilters={hasAppliedFilters}
          showNoResults={showNoResults}
        />
      )}
      {!noListings && (
        <Table
          showSorterTooltip={false}
          tableLayout="auto"
          scroll={{ x: 1752 }}
          position="center"
          dataSource={dataSource}
          className="table table--listings"
          pagination={{
            position: ['bottomLeft'],
            showSizeChanger: false,
            current: pagination.number,
            pageSize: pagination.size,
            total,
            onChange: onPaginationChangeHandler,
            hideOnSinglePage: true,
          }}
        >
          {tableColumns.map((item) => (
            <Table.Column
              {...item.props}
              key={item.key}
              title={
                item.sortKey ? (
                  <SortButton
                    onClickHandler={onSortClickHandler(item.sortKey)}
                    sortKey={item.sortKey}
                    sort={sort}
                    text={item.label}
                  />
                ) : (
                  <FormattedMessage {...item.label} />
                )
              }
              dataIndex={item.key}
            />
          ))}
          <Table.Column
            title={
              <Dropdown
                trigger="click"
                onVisibleChange={onSettingsVisibleChangeHandler}
                visible={settingsVisible}
                placement="bottomRight"
                className="table__settings"
                overlay={
                  <div className="table__settings-content">
                    <h4 className="table__settings-title mb-16">
                      <FormattedMessage id="shared.columnsSettings" />
                    </h4>
                    <Checkbox.Group
                      className="d-flex flex-column mb-0 table__settings-checkboxes"
                      value={settings}
                      onChange={onSettingsChangeHandler}
                    >
                      {LISTING_TABLE_SETTINGS.map((item) => (
                        <Checkbox
                          className={classNames({
                            'd-none': item.hiddenInSettings,
                          })}
                          key={item.key}
                          value={item.key}
                        >
                          <FormattedMessage {...item.label} />
                        </Checkbox>
                      ))}
                    </Checkbox.Group>
                  </div>
                }
              >
                <GradientButton
                  className="main-btn--icon table__settings-btn"
                  variant="tertiary"
                  addonBefore={<SvgIcon icon="settings" className="icon-right" />}
                />
              </Dropdown>
            }
            dataIndex={LISTING_TABLE_COLUMNS_KEY.SETTINGS}
            key={LISTING_TABLE_COLUMNS_KEY.SETTINGS}
            width={56}
            fixed="right"
          />
        </Table>
      )}
    </section>
  );
};

Listings.Layout = DashboardLayout;

Listings.getInitialProps = getInitialProps;

export default Listings;
