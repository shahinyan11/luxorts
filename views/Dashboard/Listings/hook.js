import { useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import { Tag } from 'antd';
import { FormattedMessage } from 'react-intl';
import classNames from 'classnames';
import { debounce } from 'lodash';

import {
  LISTING_FILTERS,
  LISTING_TABLE_COLUMNS_KEY,
  LISTING_TABLE_SETTINGS,
  LISTING_TAG_BY_STATUS,
} from 'constants/listing/manageListings';

import checkUserVerification from 'utils/auth/checkUserVerification';
import isPresent from 'utils/isPresent';

import useParametricSelector from 'hooks/useParametricSelector';
import useImmutableCallback from 'hooks/useImmutableCallback';
import useAntdDropdownVisible from 'hooks/useAntdDropdownVisible';
import useMount from 'hooks/useMount';

import {
  fetchListingBaseAmenities,
  fetchListings,
  filterListings,
  setListingsPage,
  setListingsSearchQuery,
  setListingsSortOrder,
  setListingTableSettings,
} from 'state/concepts/listings/actions';
import {
  appliedFiltersSelector,
  baseAmenitiesSelector,
  filtersSelector,
  listingsSelector,
  listingsTotalSelector,
  listingTableSettingsSelector,
  paginationSelector,
  searchQuerySelector,
  sortSelector,
} from 'state/concepts/listings/selectors';
import { fetchListingsEndpoint } from 'state/concepts/listings/endpoints';
import { loadingSelectorCreator } from 'state/data/selectors';
import { dataDeleteEntity } from 'state/data/actions';

import Settings from './Settings';
import ListingColumn from './ListingColumn';

const TIME_FORMAT = 'MMM D, YYYY';

function useContainer() {
  const dispatch = useDispatch();
  const [settingsVisible, onSettingsVisibleChangeHandler] = useAntdDropdownVisible();

  // endpoints
  const { endpoint } = fetchListingsEndpoint;

  // selectors
  const listings = useSelector(listingsSelector);
  const pagination = useSelector(paginationSelector);
  const sort = useSelector(sortSelector);
  const total = useSelector(listingsTotalSelector);
  const settings = useSelector(listingTableSettingsSelector);
  const filters = useSelector(filtersSelector);
  const searchQuery = useSelector(searchQuerySelector);
  const appliedFilters = useSelector(appliedFiltersSelector);
  const amenities = useSelector(baseAmenitiesSelector);
  const isFetching = useParametricSelector(loadingSelectorCreator, endpoint);

  // prepare data from listings
  const dataSource = useMemo(
    () =>
      listings.map((listing) => ({
        key: listing.id,
        [LISTING_TABLE_COLUMNS_KEY.LISTING]: <ListingColumn listing={listing} />,
        [LISTING_TABLE_COLUMNS_KEY.HIDDEN_TITLE]: listing.hiddenTitle || '-',
        [LISTING_TABLE_COLUMNS_KEY.GUESTS]: listing.guestsNumber || 0,
        [LISTING_TABLE_COLUMNS_KEY.BEDROOMS]: listing.bedroomsAmount || 0,
        [LISTING_TABLE_COLUMNS_KEY.BEDS]: listing.bedsAmount || 0,
        [LISTING_TABLE_COLUMNS_KEY.BATHS]: listing.bathroomsAmount || 0,
        [LISTING_TABLE_COLUMNS_KEY.LOCATION]: listing.location || '-',
        [LISTING_TABLE_COLUMNS_KEY.DATE_CREATED]: moment(listing.createdAt).format(TIME_FORMAT),
        [LISTING_TABLE_COLUMNS_KEY.LAST_UPDATED]: moment(listing.updatedAt).format(TIME_FORMAT),
        [LISTING_TABLE_COLUMNS_KEY.STATUS]: (
          <Tag className={classNames('tag', LISTING_TAG_BY_STATUS[listing.status].CLASSNAME)}>
            <FormattedMessage {...LISTING_TAG_BY_STATUS[listing.status].TITLE} />
          </Tag>
        ),
        [LISTING_TABLE_COLUMNS_KEY.SETTINGS]: (
          <Settings id={listing.id} status={listing.status} stepStatus={listing.applicationStep} />
        ),
      })),
    [listings],
  );
  const tableColumns = useMemo(
    () => LISTING_TABLE_SETTINGS.filter((item) => settings.includes(item.key)),
    [settings],
  );
  const hasAppliedFilters = isPresent(appliedFilters);

  /**
   * On sort click handler
   * @param sortKey {string}
   * @return {(function(): void)|*}
   */
  const onSortClickHandler = useImmutableCallback((sortKey) => () => {
    dispatch(setListingsSortOrder(sortKey));
    dispatch(fetchListings({ skipLoading: true }));
  });

  /**
   * On pagination change handler
   * @param pageNumber {number}
   */
  const onPaginationChangeHandler = useImmutableCallback((pageNumber) => {
    dispatch(setListingsPage(pageNumber));
    dispatch(fetchListings({ skipLoading: true }));
  });

  /**
   * On settings change handler
   * @param checkedValues {array}
   */
  const onSettingsChangeHandler = useImmutableCallback((checkedValues) => {
    dispatch(setListingTableSettings(checkedValues));
  });

  /**
   * On filter callback
   * @param selectedFilters {object}
   */
  const onFilter = useImmutableCallback((selectedFilters) => {
    dispatch(filterListings(selectedFilters));
  });

  /**
   * Clear all filters button click
   */
  const clearAllFilters = useImmutableCallback(() => {
    dispatch(filterListings(LISTING_FILTERS));
  });

  /**
   * On search callback
   * @param query {string}
   */
  const onSearch = useImmutableCallback(
    debounce((event) => {
      dispatch(setListingsSearchQuery(event.target.value || null));
      dispatch(setListingsPage(1));
      dispatch(fetchListings({ skipLoading: true }));
    }, 250),
  );

  /**
   * On component unmount handler
   */
  const onUnmountHandler = () => () => {
    dispatch(dataDeleteEntity('listing'));
  };

  /**
   * Lifecycle
   */
  useMount(onUnmountHandler);

  return {
    total,
    isFetching,
    tableColumns,
    dataSource,
    sort,
    pagination,
    settingsVisible,
    settings,
    filters,
    searchQuery,
    amenities,
    hasAppliedFilters,
    onSortClickHandler,
    onPaginationChangeHandler,
    onSettingsChangeHandler,
    onSettingsVisibleChangeHandler,
    onFilter,
    clearAllFilters,
    onSearch,
    onUnmountHandler,
  };
}

export const getInitialProps = async (ctx) => {
  const isUserVerified = checkUserVerification(ctx);

  if (!isUserVerified) {
    return;
  }

  ctx.store.dispatch(setListingsPage(1));
  ctx.store.dispatch(fetchListings());
  ctx.store.dispatch(fetchListingBaseAmenities());
};

export default useContainer;
