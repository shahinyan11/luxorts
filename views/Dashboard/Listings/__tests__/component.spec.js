import { shallow } from 'enzyme';

import {
  LISTING_FILTERS,
  LISTING_PAGINATION,
  LISTING_TABLE_COLUMNS_KEY,
} from 'constants/listing/manageListings';

import useContainer from '../hook';
import Listings from '../component';

const mockedHookData = {
  total: 1,
  isFetching: false,
  tableColumns: [
    { key: LISTING_TABLE_COLUMNS_KEY.LISTING, label: { id: 'label.id' } },
    { key: LISTING_TABLE_COLUMNS_KEY.BEDS, sortKey: 'sortKey', label: { id: 'label.id' } },
  ],
  dataSource: [],
  sort: {},
  pagination: LISTING_PAGINATION,
  settingsVisible: false,
  settings: [],
  filters: LISTING_FILTERS,
  searchQuery: 'searchQuery',
  amenities: [],
  hasAppliedFilters: false,
  onSortClickHandler: jest.fn(() => jest.fn()),
  onPaginationChangeHandler: jest.fn(),
  onSettingsChangeHandler: jest.fn(),
  onSettingsVisibleChangeHandler: jest.fn(),
  onFilter: jest.fn(),
  clearAllFilters: jest.fn(),
  onSearch: jest.fn(),
};
jest.mock('../hook', () => jest.fn(() => mockedHookData));

describe('Listings component', () => {
  let component = shallow(<Listings />);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when isFetching equals true', () => {
    useContainer.mockReturnValueOnce({
      ...mockedHookData,
      isFetching: true,
    });

    component = shallow(<Listings />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when total equals 0', () => {
    useContainer.mockReturnValueOnce({
      ...mockedHookData,
      total: 0,
    });

    component = shallow(<Listings />);

    expect(component).toMatchSnapshot();
  });
});
