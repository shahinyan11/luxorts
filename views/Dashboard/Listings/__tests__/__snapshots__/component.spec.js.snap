// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listings component matches snapshot 1`] = `
<section
  className="listing"
>
  <div
    className="listing__header mb-24"
  >
    <h1
      className="listing__title"
    >
      <Memo(FormattedMessage)
        id="shared.yourListingsCounter"
        values={
          Object {
            "count": <span
              className="listing__counter"
            >
              1
            </span>,
          }
        }
      />
    </h1>
    <Link
      href="/add-new-listing"
    >
      <a>
        <GradientButton
          addonAfter={null}
          addonBefore={null}
          className="min-width-140"
          text={
            Object {
              "id": "shared.addNewListing",
            }
          }
          variant="primary"
        />
      </a>
    </Link>
  </div>
  <Filters
    amenities={Array []}
    filters={
      Object {
        "amenities-in": Array [],
        "bathrooms_amount-eq": 0,
        "bedrooms_amount-eq": 0,
        "beds_amount-eq": 0,
        "guests_number-eq": 0,
        "status-in": Array [],
      }
    }
    onFilter={[MockFunction]}
    onSearch={[MockFunction]}
    searchQuery="searchQuery"
  />
  <ForwardRef(InternalTable)
    className="table table--listings"
    dataSource={Array []}
    pagination={
      Object {
        "current": 1,
        "hideOnSinglePage": true,
        "onChange": [MockFunction],
        "pageSize": 8,
        "position": Array [
          "bottomLeft",
        ],
        "showSizeChanger": false,
        "total": 1,
      }
    }
    position="center"
    rowKey="key"
    scroll={
      Object {
        "x": 1752,
      }
    }
    showSorterTooltip={false}
    tableLayout="auto"
  >
    <Column
      dataIndex="listing"
      key="listing"
      title={
        <Memo(FormattedMessage)
          id="label.id"
        />
      }
    />
    <Column
      dataIndex="beds"
      key="beds"
      title={
        <SortButton
          onClickHandler={[MockFunction]}
          sort={Object {}}
          sortKey="sortKey"
          text={
            Object {
              "id": "label.id",
            }
          }
        />
      }
    />
    <Column
      dataIndex="settings"
      fixed="right"
      key="settings"
      title={
        <Dropdown
          className="table__settings"
          mouseEnterDelay={0.15}
          mouseLeaveDelay={0.1}
          onVisibleChange={[MockFunction]}
          overlay={
            <div
              className="table__settings-content"
            >
              <h4
                className="table__settings-title mb-16"
              >
                <Memo(FormattedMessage)
                  id="shared.columnsSettings"
                />
              </h4>
              <Memo
                className="d-flex flex-column mb-0 table__settings-checkboxes"
                onChange={[MockFunction]}
                value={Array []}
              >
                <Checkbox
                  className="d-none"
                  value="listing"
                >
                  <Memo(FormattedMessage)
                    id="shared.listing"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="hiddenTitle"
                >
                  <Memo(FormattedMessage)
                    id="shared.hiddenTitle"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="guests"
                >
                  <Memo(FormattedMessage)
                    id="shared.guests"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="bedrooms"
                >
                  <Memo(FormattedMessage)
                    id="shared.bedrooms"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="beds"
                >
                  <Memo(FormattedMessage)
                    id="shared.beds"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="baths"
                >
                  <Memo(FormattedMessage)
                    id="shared.baths"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="location"
                >
                  <Memo(FormattedMessage)
                    id="shared.location"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="dateCreated"
                >
                  <Memo(FormattedMessage)
                    id="shared.dateCreated"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="lastUpdated"
                >
                  <Memo(FormattedMessage)
                    id="shared.lastUpdated"
                  />
                </Checkbox>
                <Checkbox
                  className=""
                  value="status"
                >
                  <Memo(FormattedMessage)
                    id="shared.status"
                  />
                </Checkbox>
              </Memo>
            </div>
          }
          placement="bottomRight"
          trigger="click"
          visible={false}
        >
          <GradientButton
            addonAfter={null}
            addonBefore={
              <SvgIcon
                className="icon-right"
                icon="settings"
              />
            }
            className="main-btn--icon table__settings-btn"
            variant="tertiary"
          />
        </Dropdown>
      }
      width={56}
    />
  </ForwardRef(InternalTable)>
</section>
`;

exports[`Listings component matches snapshot when isFetching equals true 1`] = `
<section
  className="listing"
>
  <div
    className="listing__header"
  >
    <Skeleton
      active={true}
      avatar={false}
      paragraph={true}
      title={true}
    />
  </div>
</section>
`;

exports[`Listings component matches snapshot when total equals 0 1`] = `
<section
  className="listing"
>
  <div
    className="listing__header mb-24"
  >
    <h1
      className="listing__title"
    >
      <Memo(FormattedMessage)
        id="shared.yourListingsCounter"
        values={
          Object {
            "count": <span
              className="listing__counter"
            >
              0
            </span>,
          }
        }
      />
    </h1>
    <Link
      href="/add-new-listing"
    >
      <a>
        <GradientButton
          addonAfter={null}
          addonBefore={null}
          className="min-width-140"
          text={
            Object {
              "id": "shared.addNewListing",
            }
          }
          variant="primary"
        />
      </a>
    </Link>
  </div>
  <Filters
    amenities={Array []}
    filters={
      Object {
        "amenities-in": Array [],
        "bathrooms_amount-eq": 0,
        "bedrooms_amount-eq": 0,
        "beds_amount-eq": 0,
        "guests_number-eq": 0,
        "status-in": Array [],
      }
    }
    onFilter={[MockFunction]}
    onSearch={[MockFunction]}
    searchQuery="searchQuery"
  />
  <EmptyState
    clearAllFilters={[MockFunction]}
    hasAppliedFilters={false}
    showNoResults={true}
  />
</section>
`;
